05libraries/sec02LibraryBasics.md

```diff
@@ -0,0 +1,113 @@
+---
+title: Library Basics
+---
+
+## Library Basics
+
+### Reviewing the build process
+
+When building an application there are three important steps:
+
+- **preprocessing**: follow the directives (lines started by `#`) on the files to combine the units into what's passed to the compiler;
+- **compilation**: translates the program into machine language code - object files;  and
+- **linking**: merges the various object files and links all the referred libraries as needed to create the executable.
+
+Though normally these steps are invoked by a single command, you can run them one at a time.
+
+- preprocessor:
+
+  ```bash
+  g++ -E -o <output> <input.cpp>
+  ```
+- compilation:
+
+  ```bash
+  g++ -c -o <output.o> <input.cpp>
+  ```
+- linking:
+
+  ```bash
+  g++ -o <executable> <output.o>
+  ```
+
+Directly using the compiler without a build tool (e.g., [CMake][lesson-cmake]) will eventually become too difficult and cause a mess.
+
+
+Find more details of these steps on the following resources:
+
+- [The C Preprocessor][CppAdv6] chapter on the [C++: Advanced Topics][CppAdv] course.
+- How the C++ [Compiler][CppChernoCompiler] and [Linker][CppChernoLinker] works videos by [The Cherno][Cherno].
+
+### Including libraries
+
+You need your compiler to find:
+
+* Headers: `#include`
+* Libraries:
+    * Dynamic: `.dylib` (mac), `.so` (linux), `.lib` / `.dll` (windows)
+    * Static: `.a` (\*nix), `.lib` (windows)
+
+We will see the [differences between dynamic and static libraries][lesson-DynVsSt] later. Let's see first how we include the libraries in our code.
+
+### In practice
+
+Normally, when including with `<>` the preprocessor looks for headers in the *include path list*. You can specify the include folder(s) by using the `-I` argument as needed. Similarly, the `-L` argument is used to give the path in which the linker should search for libraries to link, and the `-l` flag gives the name of the library to be linked. Note that the libraries files always start with `lib` but we don't add such prefix when referring to it.
+
+For example:
+
+```bash
+# compilation
+g++ -c -I /users/me/myproject/include main.cpp
+# linking
+g++ -o main main.o -L /users/me/myproject/lib -l mylib
+```
```

Comment:
This example would be a bit clearer if presented in the context of the project structure, and showed for example what the lib file was called (so you can explicitly see the dropped prefix). That way you can see how to go from a given project to a compile/link command rather than them reverse engineering the project structure from those commands.